modules/auxiliary/scanner/scada/modbus_zip.rb

```diff
@@ -0,0 +1,78 @@
+##
+## This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+require 'packetfu'
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Capture
+
+  def initialize(info = {})
+    super(
+      'Name'           => "Project File",
+      'Description'    => %q{
+        This module is able to extract a zip file sent through Modbus from a pcap.
+        Tested with Schneider TM221CE16R
+      },
+        'Author' => [
+          'José Diogo Monteiro <jdlopes[at]student.dei.uc.pt>',
+          'Luis Rosa <lmrosa[at]dei.uc.pt)>',
+        ],
+      'License' => MSF_LICENSE
+      )
+    register_options([Opt::RPORT(502), Opt::RHOSTS("0.0.0.0"),
+                      OptEnum.new("MODE", [true, 'Extract zip from upload/download capture', "UPLOAD",
+                                           [
+                                             "UPLOAD",
+                                             "DOWNLOAD"
+                                           ]]),
+                      OptString.new('FILENAME', [ true, "The zip file", File.join(Msf::Config.module_directory, "auxiliary", "scanner", "scada", "projectfile.zip") ]),])
+  end
+
+  def extract_zip(packet, zip_packet, first_byte, file, packet_number)
+    # ZIP start signature
+    h = packet.payload.scan(/\x50\x4B\x03\x04.*/)
+    if h.size.nonzero?
+      print_status "Zip start on packet #{packet_number + 1}"
+      file.print h[0]
+      zip_packet += 1
+      return zip_packet
+    end
+    # ZIP end signature (central directory record)
+    h = packet.payload.scan(/.*\x50\x4B\x05\x06................../)
+    if h.size.nonzero?
+      print_status "Zip end on packet #{packet_number + 1}"
+      file.print h[0][first_byte..-1]
+      zip_packet += 1
+      return zip_packet
+    end
+    # ZIP data
+    if zip_packet == 1
+      file.print packet.payload[first_byte..-1]
+    end
+    return zip_packet
+  end
+
+  def run
+    packets = PacketFu::PcapFile.read_packets datastore['PCAPFILE']
+    zip_packet = 0
+    first_byte_upload = 12
```

Comment:
first_byte_{upload,download} could be changed to constants like [this](https://github.com/rapid7/metasploit-framework/blob/master/modules/auxiliary/gather/oats_downloadservlet_traversal.rb#L69).